File: meddraft_ai/prompts/medical_research_skills/scientific-skills/Data Analysis/meta-sensitivity-plot/scripts/sensitivity_analysis.R
#!/usr/bin/env Rscript
# 敏感性分析脚本（逐一剔除法）- 含图片输出
# 用法: Rscript sensitivity_analysis.R <csv_path> <type> [outcome_name] [output_dir]

suppressPackageStartupMessages({
  library(meta)
  library(metafor)
  library(stringr)
  library(grid)
})

# 解析命令行参数
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  stop("用法: Rscript sensitivity_analysis.R <csv_path> <type> [outcome_name] [output_dir]")
}

csv_path <- args[1]
type <- args[2]
outcome_name <- if (length(args) >= 3 && args[3] != "") args[3] else NULL
output_dir <- if (length(args) >= 4 && args[4] != "") args[4] else dirname(csv_path)

# 验证类型
if (!type %in% c("Binary", "Continuity", "Survival")) {
  stop("type 必须是 Binary, Continuity 或 Survival")
}

# 读取数据
if (!file.exists(csv_path)) {
  stop(paste0("文件不存在: ", csv_path))
}

data_df <- read.csv(csv_path, stringsAsFactors = FALSE)

# 获取结局指标名称
if (is.null(outcome_name)) {
  if ("outcome_new" %in% colnames(data_df)) {
    outcome_name <- unique(data_df$outcome_new)[1]
  } else {
    outcome_name <- "Outcome"
  }
}

# 验证数据量
if (nrow(data_df) < 3) {
  stop("至少需要3项研究才能进行敏感性分析")
}

# 创建输出目录
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

# 辅助函数：获取列
get_col <- function(df, candidates) {
  for (col in candidates) {
    if (col %in% colnames(df)) return(df[[col]])
  }
  return(NULL)
}

# 根据类型创建meta分析对象
if (type == "Binary") {
  required_cols <- c("study", "group1_Events", "group1_sample_size",
                     "group2_Events", "group2_sample_size")
  missing_cols <- setdiff(required_cols, colnames(data_df))
  if (length(missing_cols) > 0) {
    stop(paste0("缺少必要列: ", paste(missing_cols, collapse = ", ")))
  }

  meta_analysis <- metabin(
    event.e = as.numeric(data_df$group1_Events),
    n.e = as.numeric(data_df$group1_sample_size),
    event.c = as.numeric(data_df$group2_Events),
    n.c = as.numeric(data_df$group2_sample_size),
    studlab = data_df$study,
    sm = "OR",
    method = "MH",
    random = TRUE
  )
  effect_name <- "OR"
  effect_label <- "Odds Ratio"

} else if (type == "Continuity") {
  required_cols <- c("study", "group1_sample_size", "group1_Mean", "group1_SD",
                     "group2_sample_size", "group2_Mean", "group2_SD")
  missing_cols <- setdiff(required_cols, colnames(data_df))
  if (length(missing_cols) > 0) {
    stop(paste0("缺少必要列: ", paste(missing_cols, collapse = ", ")))
  }

  meta_analysis <- metacont(
    studlab = data_df$study,
    n.e = as.numeric(data_df$group1_sample_size),
    mean.e = as.numeric(data_df$group1_Mean),
    sd.e = as.numeric(data_df$group1_SD),
    n.c = as.numeric(data_df$group2_sample_size),
    mean.c = as.numeric(data_df$group2_Mean),
    sd.c = as.numeric(data_df$group2_SD),
    sm = "SMD",
    method.tau = "DL",
    method.random.ci = "HK"
  )
  effect_name <- "SMD"
  effect_label <- "Standardized Mean Difference"

} else if (type == "Survival") {
  hr_col <- get_col(data_df, c("group1_HR"))
  lower_col <- get_col(data_df, c("group1_95%Lower CI", "group1_95.Lower.CI"))
  upper_col <- get_col(data_df, c("group1_95%Upper CI", "group1_95.Upper.CI"))

  if (is.null(hr_col) || is.null(lower_col) || is.null(upper_col)) {
    stop("缺少必要列: group1_HR, group1_95%Lower CI, group1_95%Upper CI")
  }

  hr <- as.numeric(hr_col)
  lower_ci <- as.numeric(lower_col)
  upper_ci <- as.numeric(upper_col)

  valid_idx <- !is.na(hr) & !is.na(lower_ci) & !is.na(upper_ci) &
    hr > 0 & lower_ci > 0 & upper_ci > 0

  if (sum(valid_idx) < 3) {
    stop("有效数据不足3项，无法进行敏感性分析")
  }

  log_hr <- log(hr[valid_idx])
  se <- (log(upper_ci[valid_idx]) - log(lower_ci[valid_idx])) / (2 * 1.96)

  meta_analysis <- metagen(
    TE = log_hr,
    seTE = se,
    studlab = data_df$study[valid_idx],
    sm = "HR",
    method.tau = "DL",
    common = TRUE
  )
  effect_name <- "HR"
  effect_label <- "Hazard Ratio"
}

# 敏感性分析 - 逐一排除法
Leave_one_fun <- tryCatch({
  metainf(meta_analysis, pooled = "random")
}, error = function(e) {
  stop(paste0("敏感性分析失败: ", e$message))
})

# ========== 绘制敏感性分析森林图 ==========
plot_file <- file.path(output_dir, paste0(type, "_sensitive_forest_", outcome_name, ".png"))

# 计算图片高度（根据研究数量动态调整）
n_studies <- length(Leave_one_fun$studlab)
plot_height <- max(600, 150 + n_studies * 35)

png(plot_file, width = 900, height = plot_height, units = 'px', pointsize = 14)
tryCatch({
  forest(Leave_one_fun)
}, error = function(e) {
  # 如果forest失败，使用简单的散点图替代
  n <- length(Leave_one_fun$studlab)
  te <- if(effect_name %in% c("OR", "HR")) exp(Leave_one_fun$TE) else Leave_one_fun$TE
  ci_lower <- if(effect_name %in% c("OR", "HR")) exp(Leave_one_fun$lower) else Leave_one_fun$lower
  ci_upper <- if(effect_name %in% c("OR", "HR")) exp(Leave_one_fun$upper) else Leave_one_fun$upper

  par(mar = c(5, 10, 4, 2))
  plot(te, 1:n, xlim = range(c(ci_lower, ci_upper), na.rm = TRUE),
       ylim = c(0.5, n + 0.5), pch = 15, col = "darkblue",
       xlab = effect_label, ylab = "", yaxt = "n",
       main = paste0("Leave-one-out ", effect_label))
  axis(2, at = 1:n, labels = Leave_one_fun$studlab, las = 1)
  segments(ci_lower, 1:n, ci_upper, 1:n, col = "darkblue")
  abline(v = if(effect_name %in% c("OR", "HR")) 1 else 0, lty = 2, col = "gray")
})
dev.off()

# ========== 绘制漏斗图（Funnel plot） ==========
funnel_file <- file.path(output_dir, paste0(type, "_funnel_", outcome_name, ".png"))

png(funnel_file, width = 700, height = 700, units = 'px', pointsize = 14)
tryCatch({
  # 使用原始meta_analysis对象绘制漏斗图
  if (exists("meta_analysis") && !is.null(meta_analysis)) {
    # 对于二分类和生存数据，使用meta对象直接绘图
    funnel(meta_analysis, studlab = TRUE, xlab = effect_label, contour = c(0.90, 0.95), col.contour = c("gray80","gray60"), legend = TRUE)
  } else {
    # 作为降级选项，绘制基于效应量和标准误的散点图
    te <- Leave_one_fun$TE
    se <- Leave_one_fun$se
    plot(te, se, pch = 16, xlab = effect_label, ylab = "Standard Error", main = paste0("Funnel plot - ", outcome_name))
  }
}, error = function(e) {
  # 若funnel绘制失败，输出简单提示图
  plot.new()
  text(0.5, 0.5, paste0("Funnel plot failed:\n", e$message), cex = 0.9)
})
dev.off()

# ========== 提取结果数据 ==========
# 提取效应量用于显示
if (effect_name %in% c("OR", "HR")) {
  te_display <- exp(Leave_one_fun$TE)
  lower_display <- exp(Leave_one_fun$lower)
  upper_display <- exp(Leave_one_fun$upper)
  pooled_display <- exp(meta_analysis$TE.random)
  pooled_lower <- exp(meta_analysis$lower.random)
  pooled_upper <- exp(meta_analysis$upper.random)
} else {
  te_display <- Leave_one_fun$TE
  lower_display <- Leave_one_fun$lower
  upper_display <- Leave_one_fun$upper
  pooled_display <- meta_analysis$TE.random
  pooled_lower <- meta_analysis$lower.random
  pooled_upper <- meta_analysis$upper.random
}

# 格式化数值
format_val <- function(x, digits = 2) {
  ifelse(is.na(x) | is.nan(x) | is.null(x), "NA", format(round(x, digits), nsmall = digits))
}

sensitivity_data <- data.frame(
  study = Leave_one_fun$studlab,
  value = format_val(te_display, 2),
  lower_95 = format_val(lower_display, 2),
  upper_95 = format_val(upper_display, 2),
  statistic = Leave_one_fun$statistic,
  p_val = Leave_one_fun$pval,
  tau2 = Leave_one_fun$tau2,
  I2 = Leave_one_fun$I2,
  stringsAsFactors = FALSE
)

# 保存结果
csv_file <- file.path(output_dir, paste0(type, "_sensitive_", outcome_name, ".csv"))
write.csv(sensitivity_data, csv_file, row.names = FALSE)

# 判断稳健性
effect_values <- as.numeric(sensitivity_data$value)
pooled_effect_val <- effect_values[length(effect_values)]

# 检查所有剔除后的效应量方向是否一致
if (effect_name %in% c("OR", "HR")) {
  all_greater <- all(effect_values > 1, na.rm = TRUE)
  all_smaller <- all(effect_values < 1, na.rm = TRUE)
  all_same_direction <- all_greater | all_smaller
} else {
  all_greater <- all(effect_values > 0, na.rm = TRUE)
  all_smaller <- all(effect_values < 0, na.rm = TRUE)
  all_same_direction <- all_greater | all_smaller
}

robustness <- if (all_same_direction) "稳健" else "不稳健（剔除部分研究后效应量方向改变）"

# 计算效应量变化范围
effect_range <- range(effect_values, na.rm = TRUE)
effect_variation <- round((effect_range[2] - effect_range[1]) / abs(pooled_effect_val) * 100, 1)

# 输出结果摘要
cat("\n")
cat("═══════════════════════════════════════════\n")
cat("敏感性分析完成\n")
cat("═══════════════════════════════════════════\n\n")
cat(paste0("【结局指标】", outcome_name, "\n"))
cat(paste0("【数据类型】", type, "\n"))
cat(paste0("【纳入研究】", nrow(data_df), " 项\n\n"))

cat("【输出文件】\n")
cat(paste0("• 敏感性分析森林图：", plot_file, "\n"))
cat(paste0("• 敏感性分析数据表：", csv_file, "\n\n"))

cat("【合并效应量（全部研究）】\n")
cat(paste0("• ", effect_name, " = ", round(pooled_display, 2),
           " [", round(pooled_lower, 2), "; ", round(pooled_upper, 2), "]\n\n"))

cat("【敏感性分析结果摘要】\n")
cat(sprintf("%-20s %-10s %-18s %-8s\n", "剔除研究", effect_name, "95% CI", "I²"))
cat("───────────────────────────────────────────\n")

for (i in 1:nrow(sensitivity_data)) {
  study_name <- if (nchar(sensitivity_data$study[i]) > 18) {
    paste0(substr(sensitivity_data$study[i], 1, 15), "...")
  } else {
    sensitivity_data$study[i]
  }
  ci_str <- paste0("[", sensitivity_data$lower_95[i], "; ", sensitivity_data$upper_95[i], "]")
  i2_val <- sensitivity_data$I2[i]
  if (is.na(i2_val) || is.nan(i2_val) || is.null(i2_val)) {
    i2_str <- "NA"
  } else {
    i2_str <- paste0(round(as.numeric(i2_val) * 100, 1), "%")
  }
  cat(sprintf("%-20s %-10s %-18s %-8s\n",
              study_name, sensitivity_data$value[i], ci_str, i2_str))
}

cat("\n【效应量变化分析】\n")
cat(paste0("• 效应量范围：", round(effect_range[1], 2), " ~ ", round(effect_range[2], 2), "\n"))
cat(paste0("• 相对变化幅度：", effect_variation, "%\n\n"))

cat("【结论】\n")
cat(paste0("• 效应量稳健性：", robustness, "\n"))

if (effect_variation < 10) {
  cat("• 剔除任意单项研究后效应量变化小（<10%），结果稳健。\n")
} else if (effect_variation < 20) {
  cat("• 剔除部分研究后效应量有一定变化（10-20%），需关注影响较大的研究。\n")
} else {
  cat("• 效应量变化较大（>20%），部分研究对结果影响显著，建议进一步分析。\n")
}

cat("═══════════════════════════════════════════\n")

File: meddraft_ai/prompts/medical_research_skills/scientific-skills/Data Analysis/meta-sensitivity-plot/scripts/sensitivity_analysis.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Sensitivity analysis script (one-by-one elimination method) - Python alternative version
Usage: python sensitivity_analysis.py <csv_path> <type> [outcome_name] [output_dir]

When the R script drawing fails, use this Python script as a backup solution for sensitivity analysis and drawing."""

import sys
import os
import argparse
import warnings
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from scipy import stats
from datetime import datetime

warnings.filterwarnings('ignore')

# Set Chinese font
plt.rcParams['font.sans-serif'] = ['SimHei', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False


class MetaAnalyzer:
    """Meta-analysis sensitivity analyzer"""
    
    def __init__(self, data_df, data_type, outcome_name="Outcome"):
        self.data_df = data_df
        self.data_type = data_type
        self.outcome_name = outcome_name
        self.n_studies = len(data_df)
        
        if self.n_studies < 3:
            raise ValueError("At least 3 studies are required to conduct sensitivity analysis")
    
    def validate_data(self):
        """Verify data integrity"""
        required_cols = {
            'Binary': ['study', 'group1_Events', 'group1_sample_size', 
                      'group2_Events', 'group2_sample_size'],
            'Continuity': ['study', 'group1_sample_size', 'group1_Mean', 'group1_SD',
                          'group2_sample_size', 'group2_Mean', 'group2_SD'],
            'Survival': ['study', 'group1_HR', 'group1_95%Lower CI', 'group1_95%Upper CI']
        }
        
        cols = required_cols.get(self.data_type, [])
        missing = [c for c in cols if c not in self.data_df.columns]
        
        if missing:
            raise ValueError(f"Missing required column: {', '.join(missing)}")
        
        return True
    
    def calculate_ci(self, te, se, level=0.95):
        """Calculate confidence intervals"""
        z = stats.norm.ppf((1 + level) / 2)
        return te - z * se, te + z * se
    
    def meta_binary(self):
        """Two-category meta-analysis"""
        data = self.data_df.copy()
        data['e1'] = pd.to_numeric(data['group1_Events'], errors='coerce')
        data['n1'] = pd.to_numeric(data['group1_sample_size'], errors='coerce')
        data['e2'] = pd.to_numeric(data['group2_Events'], errors='coerce')
        data['n2'] = pd.to_numeric(data['group2_sample_size'], errors='coerce')
        
        # Mantel-Haenszel estimate
        results = []
        for idx, row in data.iterrows():
            if pd.isna([row['e1'], row['n1'], row['e2'], row['n2']]).any():
                continue
            
            # Use log OR
            p1 = (row['e1'] + 0.5) / (row['n1'] + 1)
            p2 = (row['e2'] + 0.5) / (row['n2'] + 1)
            te = np.log(p1 / p2 * (1 - p2) / (1 - p1))
            se = np.sqrt(1 / (row['e1'] + 0.5) + 1 / (row['n1'] - row['e1'] + 0.5) +
                        1 / (row['e2'] + 0.5) + 1 / (row['n2'] - row['e2'] + 0.5))
            
            results.append({
                'study': row['study'],
                'te': te,
                'se': se,
                'n': row['n1'] + row['n2']
            })
        
        return pd.DataFrame(results), 'OR'
    
    def meta_continuity(self):
        """Continuous variable meta-analysis"""
        data = self.data_df.copy()
        
        results = []
        for idx, row in data.iterrows():
            n1 = pd.to_numeric(row['group1_sample_size'], errors='coerce')
            m1 = pd.to_numeric(row['group1_Mean'], errors='coerce')
            sd1 = pd.to_numeric(row['group1_SD'], errors='coerce')
            n2 = pd.to_numeric(row['group2_sample_size'], errors='coerce')
            m2 = pd.to_numeric(row['group2_Mean'], errors='coerce')
            sd2 = pd.to_numeric(row['group2_SD'], errors='coerce')
            
            if pd.isna([n1, m1, sd1, n2, m2, sd2]).any():
                continue
            
            # Calculate the standardized mean difference (SMD)
            pooled_sd = np.sqrt(((n1 - 1) * sd1**2 + (n2 - 1) * sd2**2) / (n1 + n2 - 2))
            te = (m1 - m2) / pooled_sd if pooled_sd > 0 else 0
            se = np.sqrt(1/n1 + 1/n2 + te**2 / (2 * (n1 + n2)))
            
            results.append({
                'study': row['study'],
                'te': te,
                'se': se,
                'n': n1 + n2
            })
        
        return pd.DataFrame(results), 'SMD'
    
    def meta_survival(self):
        """Survival meta-analysis"""
        data = self.data_df.copy()
        
        results = []
        for idx, row in data.iterrows():
            hr = pd.to_numeric(row['group1_HR'], errors='coerce')
            lower_ci = pd.to_numeric(row.get('group1_95%Lower CI', None) or 
                                    row.get('group1_95.Lower.CI', None), errors='coerce')
            upper_ci = pd.to_numeric(row.get('group1_95%Upper CI', None) or 
                                    row.get('group1_95.Upper.CI', None), errors='coerce')
            
            if pd.isna([hr, lower_ci, upper_ci]).any() or hr <= 0 or lower_ci <= 0 or upper_ci <= 0:
                continue
            
            # Log conversion
            te = np.log(hr)
            se = (np.log(upper_ci) - np.log(lower_ci)) / (2 * 1.96)
            
            results.append({
                'study': row['study'],
                'te': te,
                'se': se,
                'n': 100  # Assuming sample size
            })
        
        if len(results) < 3:
            raise ValueError("There are less than 3 valid data items and sensitivity analysis cannot be performed.")
        
        return pd.DataFrame(results), 'HR'
    
    def inverse_variance_meta(self, meta_df):
        """Meta-analysis using inverse variance pooling method"""
        meta_df = meta_df.copy()
        meta_df['weight'] = 1 / (meta_df['se'] ** 2)
        total_weight = meta_df['weight'].sum()
        
        te_pooled = (meta_df['te'] * meta_df['weight']).sum() / total_weight
        se_pooled = 1 / np.sqrt(total_weight)
        
        # Compute heterogeneity
        q = ((meta_df['te'] - te_pooled) ** 2 / (meta_df['se'] ** 2)).sum()
        df = len(meta_df) - 1
        i2 = max(0, (q - df) / q) if q > df else 0
        
        return {
            'te': te_pooled,
            'se': se_pooled,
            'q': q,
            'i2': i2,
            'p_value': 1 - stats.chi2.cdf(q, df) if df > 0 else 1.0
        }
    
    def sensitivity_analysis(self, meta_results, effect_label):
        """Perform sensitivity analysis (one-by-one elimination method)"""
        sensitivity_list = []
        
        for leave_out_idx in range(len(meta_results)):
            # Construct a dataset with the i-th study removed
            subset = meta_results.drop(leave_out_idx).reset_index(drop=True)
            
            if len(subset) < 2:
                continue
            
            # Meta-analysis
            pooled = self.inverse_variance_meta(subset)
            
            study_name = meta_results.iloc[leave_out_idx]['study']
            study_name = study_name[:18] + '...' if len(study_name) > 18 else study_name
            
            sensitivity_list.append({
                'study_leave_out': study_name,
                'te': pooled['te'],
                'se': pooled['se'],
                'i2': pooled['i2'],
                'p_value': pooled['p_value'],
                'q': pooled['q']
            })
        
        return pd.DataFrame(sensitivity_list)
    
    def format_effect(self, te, effect_label):
        """Formatting effect sizes based on effect metric type"""
        if effect_label in ['OR', 'HR']:
            return np.exp(te)
        return te
    
    def analyze(self):
        """Perform a complete sensitivity analysis"""
        self.validate_data()
        
        # Select analysis method based on type
        if self.data_type == 'Binary':
            meta_results, effect_label = self.meta_binary()
        elif self.data_type == 'Continuity':
            meta_results, effect_label = self.meta_continuity()
        elif self.data_type == 'Survival':
            meta_results, effect_label = self.meta_survival()
        else:
            raise ValueError(f"Unsupported data type: {self.data_type}")
        
        if len(meta_results) < 3:
            raise ValueError("There are less than 3 valid data items and sensitivity analysis cannot be performed.")
        
        # Meta-analysis of all studies
        pooled_all = self.inverse_variance_meta(meta_results)
        
        # sensitivity analysis
        sensitivity_df = self.sensitivity_analysis(meta_results, effect_label)
        
        # Format results for output
        te_display = self.format_effect(sensitivity_df['te'], effect_label)
        lower, upper = self.calculate_ci(sensitivity_df['te'], sensitivity_df['se'])
        lower_display = self.format_effect(lower, effect_label)
        upper_display = self.format_effect(upper, effect_label)
        
        pooled_display = self.format_effect(pooled_all['te'], effect_label)
        pooled_lower, pooled_upper = self.calculate_ci(pooled_all['te'], pooled_all['se'])
        pooled_lower = self.format_effect(pooled_lower, effect_label)
        pooled_upper = self.format_effect(pooled_upper, effect_label)
        
        return {
            'meta_results': meta_results,
            'sensitivity_df': sensitivity_df,
            'te_display': te_display,
            'lower_display': lower_display,
            'upper_display': upper_display,
            'pooled_display': pooled_display,
            'pooled_lower': pooled_lower,
            'pooled_upper': pooled_upper,
            'pooled_all': pooled_all,
            'effect_label': effect_label
        }


def plot_forest(sensitivity_df, meta_results, effect_label, outcome_name, output_file):
    """Plot a sensitivity analysis forest plot"""
    n_studies = len(sensitivity_df)
    plot_height = max(6, 3 + n_studies * 0.3)
    
    fig, ax = plt.subplots(figsize=(10, plot_height))
    
    # Extract data
    studies = sensitivity_df['study_leave_out'].values
    te = sensitivity_df['te'].values
    se = sensitivity_df['se'].values
    
    # Convert display value
    if effect_label in ['OR', 'HR']:
        display_te = np.exp(te)
        ci_lower, ci_upper = np.exp(te - 1.96 * se), np.exp(te + 1.96 * se)
    else:
        display_te = te
        ci_lower, ci_upper = te - 1.96 * se, te + 1.96 * se
    
    # draw
    y_pos = np.arange(n_studies, 0, -1)
    
    for i, (y, te_val, lower, upper) in enumerate(zip(y_pos, display_te, ci_lower, ci_upper)):
        # caliber line
        ax.plot([lower, upper], [y, y], 'b-', linewidth=1.5)
        # point of effect
        ax.plot(te_val, y, 'o', color='darkblue', markersize=8)
    
    # reference line
    if effect_label in ['OR', 'HR']:
        ax.axvline(x=1, color='gray', linestyle='--', linewidth=1, alpha=0.7)
    else:
        ax.axvline(x=0, color='gray', linestyle='--', linewidth=1, alpha=0.7)
    
    ax.set_yticks(y_pos)
    ax.set_yticklabels(studies)
    ax.set_xlabel(effect_label, fontsize=12)
    ax.set_title(f'Leave-one-out {effect_label} ({outcome_name})', fontsize=14, fontweight='bold')
    ax.set_ylim(0, n_studies + 1)
    ax.grid(axis='x', alpha=0.3)
    
    plt.tight_layout()
    try:
        plt.savefig(output_file, dpi=100, bbox_inches='tight', encoding='utf-8')
        print(f"✓ Sensitivity analysis forest plot: {output_file}")
    except Exception as e:
        print(f"✗ Forest map failed to save: {e}")
    plt.close()


def plot_funnel(meta_results, effect_label, outcome_name, output_file):
    """Draw a funnel chart"""
    fig, ax = plt.subplots(figsize=(8, 6))
    
    te = meta_results['te'].values
    se = meta_results['se'].values
    
    # Convert display value
    if effect_label in ['OR', 'HR']:
        display_te = np.exp(te)
    else:
        display_te = te
    
    ax.scatter(display_te, se, alpha=0.6, s=100, color='darkblue')
    
    # Draw lines of symmetry
    max_se = se.max() * 1.1
    if effect_label in ['OR', 'HR']:
        reference = 1
    else:
        reference = 0
    
    se_line = np.linspace(0, max_se, 100)
    for z in [1.96, 3.29]:
        ax.plot(reference + z * se_line, se_line, 'k--', alpha=0.3, linewidth=1)
        ax.plot(reference - z * se_line, se_line, 'k--', alpha=0.3, linewidth=1)
    
    ax.axvline(x=reference, color='red', linestyle='-', linewidth=2, alpha=0.5)
    ax.set_xlabel(effect_label, fontsize=12)
    ax.set_ylabel('Standard Error', fontsize=12)
    ax.set_title(f'Funnel Plot ({outcome_name})', fontsize=14, fontweight='bold')
    ax.invert_yaxis()
    ax.grid(alpha=0.3)
    
    # Add tag
    for i, (x, y) in enumerate(zip(display_te, se)):
        study_name = meta_results.iloc[i]['study']
        if len(study_name) > 10:
            study_name = study_name[:7] + '...'
        ax.annotate(study_name, (x, y), fontsize=8, alpha=0.7, 
                   xytext=(5, 5), textcoords='offset points')
    
    plt.tight_layout()
    try:
        plt.savefig(output_file, dpi=100, bbox_inches='tight', encoding='utf-8')
        print(f"✓ funnel chart: {output_file}")
    except Exception as e:
        print(f"✗ Funnel chart saving failed: {e}")
    plt.close()


def save_results_csv(results, output_file):
    """Save sensitivity analysis results to CSV"""
    output_df = pd.DataFrame({
        'study': results['sensitivity_df']['study_leave_out'],
        'value': results['te_display'].round(2),
        'lower_95': results['lower_display'].round(2),
        'upper_95': results['upper_display'].round(2),
        'I2': (results['sensitivity_df']['i2'] * 100).round(1),
        'p_value': results['sensitivity_df']['p_value'].round(4),
        'Q_statistic': results['sensitivity_df']['q'].round(2)
    })
    
    output_df.to_csv(output_file, index=False, encoding='utf-8')
    print(f"✓ Sensitivity Analysis Data Sheet: {output_file}")


def print_summary(results, data_type, outcome_name, n_studies):
    """Print result summary"""
    effect_label = results['effect_label']
    
    # Judgment Robustness
    effect_values = results['te_display'].values
    pooled_val = results['pooled_display']
    
    if effect_label in ['OR', 'HR']:
        all_greater = np.all(effect_values > 1)
        all_smaller = np.all(effect_values < 1)
    else:
        all_greater = np.all(effect_values > 0)
        all_smaller = np.all(effect_values < 0)
    
    robustness = 'steady' if (all_greater or all_smaller) else 'Not robust (the direction of the effect size changes after removing some studies)'
    
    effect_range = [np.nanmin(effect_values), np.nanmax(effect_values)]
    effect_variation = abs((effect_range[1] - effect_range[0]) / pooled_val * 100) if pooled_val != 0 else 0
    
    # output
    print("\n" + "="*50)
    print("Sensitivity analysis completed")
    print("="*50 + "\n")
    print(f"【Outcome indicators】{outcome_name}")
    print(f"【data type】{data_type}")
    print(f"【Included studies】{n_studies} item\n")
    
    print("[Pooled effect size (all studies)]")
    print(f"• {effect_label} = {results['pooled_display']:.2f} " +
          f"[{results['pooled_lower']:.2f}; {results['pooled_upper']:.2f}]\n")
    
    print("[Summary of sensitivity analysis results]")
    print(f"{'Eliminate studies':<20} {effect_label:<10} {'95% CI':<20} {'I²':<8}")
    print("-"*50)
    
    for i, row in results['sensitivity_df'].iterrows():
        study = row['study_leave_out']
        value = f"{results['te_display'].iloc[i]:.2f}"
        ci = f"[{results['lower_display'].iloc[i]:.2f}; {results['upper_display'].iloc[i]:.2f}]"
        i2 = f"{row['i2']*100:.1f}%" if not np.isnan(row['i2']) else "NA"
        print(f"{study:<20} {value:<10} {ci:<20} {i2:<8}")
    
    print("【Effect size change analysis】")
    print(f"• effect size range：{effect_range[0]:.2f} ~ {effect_range[1]:.2f}")
    print(f"• Relative change amplitude：{effect_variation:.1f}%\n")
    
    print("【in conclusion】")
    print(f"• Effect size robustness：{robustness}")
    
    if effect_variation < 10:
        print("• The effect size change is small (<10%) after excluding any single study, and the results are robust.")
    elif effect_variation < 20:
        print("• After excluding some studies, the effect size will change to a certain extent (10-20%), so attention should be paid to studies with greater impact.")
    else:
        print("• The effect size varies greatly (>20%), and some studies have a significant impact on the results. Further analysis is recommended.")
    
    print("="*50 + "\n")


def main():
    # Parse command line parameters
    parser = argparse.ArgumentParser(description='Meta-analysis sensitivity analysis (Python alternative version)')
    parser.add_argument('csv_path', help='Enter CSV file path')
    parser.add_argument('type', choices=['Binary', 'Continuity', 'Survival'],
                       help='data type')
    parser.add_argument('outcome_name', nargs='?', default='Outcome',
                       help='Outcome indicator name (optional)')
    parser.add_argument('output_dir', nargs='?', default=None,
                       help='Output directory (optional)')
    
    args = parser.parse_args()
    
    csv_path = args.csv_path
    data_type = args.type
    outcome_name = args.outcome_name if args.outcome_name else 'Outcome'
    output_dir = args.output_dir if args.output_dir else os.path.dirname(csv_path)
    
    # Validate input files
    if not os.path.exists(csv_path):
        print(f"✗ File does not exist: {csv_path}")
        sys.exit(1)
    
    # Create output directory
    if not os.path.exists(output_dir):
        os.makedirs(output_dir, exist_ok=True)
    
    # Read data
    try:
        data_df = pd.read_csv(csv_path)
    except Exception as e:
        print(f"✗ Unable to readCSVdocument: {e}")
        sys.exit(1)
    
    # Perform analysis
    try:
        analyzer = MetaAnalyzer(data_df, data_type, outcome_name)
        results = analyzer.analyze()
        
        # Save results
        csv_output = os.path.join(output_dir, f"{data_type}_sensitive_{outcome_name}.csv")
        forest_output = os.path.join(output_dir, f"{data_type}_sensitive_forest_{outcome_name}.png")
        funnel_output = os.path.join(output_dir, f"{data_type}_funnel_{outcome_name}.png")
        
        save_results_csv(results, csv_output)
        plot_forest(results['sensitivity_df'], results['meta_results'], 
                   results['effect_label'], outcome_name, forest_output)
        plot_funnel(results['meta_results'], results['effect_label'], 
                   outcome_name, funnel_output)
        
        # Print summary
        print_summary(results, data_type, outcome_name, len(data_df))
        
        print("✓ Python sensitivity analysis (alternative plan) was executed successfully!")
        
    except Exception as e:
        print(f"✗ Analysis failed: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == '__main__':
    main()
